torch/utils/data/_utils/stateful.py

```diff
@@ -0,0 +1,50 @@
+# mypy: allow-untyped-defs
+"""
+Stateful protocol for objects that can save and restore their state.
+
+This module defines the Stateful protocol that objects can implement to 
+participate in DataLoader checkpointing when using stateful=True.
+"""
+
+from typing import Any, Dict, Protocol, runtime_checkable
+
+
+@runtime_checkable
+class Stateful(Protocol):
```

Comment:
TIL that this protocol already exists in core: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/stateful.py.  

Should we use it from here instead of redefining it ourselves?  Or perhaps the definition should be moved to a more central location (e.g. something like torch/utils/stateful.py)